mmpy_bot/plugins/base.py

```diff
@@ -88,21 +71,185 @@ async def call_function(
+        """Shows this help information."""
+        self.driver.reply_to(message, self.get_help_string(), direct=self.direct_help)
+
+
+@dataclass
+class PluginHelpInfo:
+    help_type: str
+    location: str
+    function: Function
+    pattern: str
+    plugin_docheader: str
+    plugin_docfull: str
+    function_docheader: str
+    function_docfull: str
+    direct: bool
+    mention: bool
+    is_click: bool
+    annotations: Dict
+
+
+class PluginManager:
+    """PluginManager is responsible for initializing all plugins and display aggregated
+    help from each of them.
+
+    It is supposed to be transparent to EventHandler that interacts directly with each
+    individual Plugin.
+    """
+
+    def __init__(
+        self,
+        plugins: Sequence[Plugin],
+    ):
+        self.settings: Optional[Settings] = None
+        self.plugins: Sequence[Plugin] = plugins
+
+        self.message_listeners: Dict[re.Pattern, List[MessageFunction]] = defaultdict(
+            list
+        )
+        self.webhook_listeners: Dict[re.Pattern, List[WebHookFunction]] = defaultdict(
+            list
+        )
+
+    def __iter__(self):
+        return iter(self.plugins)
+
+    def initialize_manager(self, driver: Driver, settings: Settings):
+        for plugin in self.plugins:
+            plugin.initialize(driver, self, settings)
+
+            # Register listeners for any listener functions in the plugin
+            for attribute in dir(plugin):
+                attribute = getattr(plugin, attribute)
+                if not isinstance(attribute, Function):
+                    continue
+
+                # Register this function and any potential siblings
+                for function in [attribute] + attribute.siblings:
+                    # Plugin message/webhook handlers can be decorated multiple times
+                    # resulting in multiple siblings that do not have .plugin defined
+                    # or where the relationship with the parent plugin is incorrect
+                    function.plugin = plugin
+                    if isinstance(function, MessageFunction):
+                        self.message_listeners[function.matcher].append(function)
+                    elif isinstance(function, WebHookFunction):
+                        self.webhook_listeners[function.matcher].append(function)
+                    else:
+                        raise TypeError(
+                            f"{plugin.__class__.__name__} has a function of unsupported"
+                            f" type {type(function)}."
+                        )
+
+    def _split_docstring(self, doc):
+        """Split docstring into first line (header) and full body."""
+        return (doc.split("\n", 1)[0], doc) if doc is not None else ("", "")
+
+    def _generate_plugin_help(
+        self,
+        plug_help: List[PluginHelpInfo],
+        help_type: str,
+        items: ItemsView[re.Pattern, List[Function]],
```

Comment:
Just a personal preference, but I think it's cleaner to just loop over the original dict here rather than specifically requiring an `ItemsView`. My default assumption is that a function does not modify its arguments in place.